### PR TITLE
Schema generation

### PR DESCRIPTION
**Pull request by minasmart**
#This generates a schema that's consumable by a query builder thingy. It allows for dynamic fields and relationships, and gives you early warnings about things changing on the server side.

The product schema thing looks like this now.

```javascript
{
  "name": "Product",
  "fields": [
    "createdAt",
    "handle",
    "id",
    "productType",
    "publishedAt",
    "tags",
    "title",
    "updatedAt",
    "vendor"
  ],
  "fieldsWithArgs": {},
  "relationships": {
    "collections": {
      "type": "CollectionConnection",
      "args": [
        "first",
        "after",
        "reverse"
      ],
      "schemaModule": "collection-connection"
    },
    "images": {
      "type": "Image",
      "args": [
        "first",
        "maxWidth",
        "maxHeight",
        "crop",
        "scale"
      ],
      "schemaModule": "image"
    },
    "options": {
      "type": "ProductOption",
      "args": [
        "first"
      ],
      "schemaModule": "product-option"
    },
    "variants": {
      "type": "ProductVariant",
      "args": [
        "first"
      ],
      "schemaModule": "product-variant"
    }
  },
  "fieldOf": [
    {
      "type": "QueryRoot",
      "schemaModule": "query-root"
    },
    {
      "type": "Collect",
      "schemaModule": "collect"
    },
    {
      "type": "ProductEdge",
      "schemaModule": "product-edge"
    }
  ]
}
```

__NOTE:__ This is only for experimentation within the `graphql` branch. Not being merged into master.

### CODE REVIEW

**Comment by minasmart** [comment]
@tessalt @lreeves have a look?

File: build-lib/lib.js [hunk 1]
```diff
@@ -23,7 +23,13 @@ function sourceTree(pathConfig, moduleType) {
     babelConfig(null, moduleType)
   );
 
-  return mergeTrees([lib, shims]);
+  const schema = babelTranspiler(
+    (new GraphSchema()),
+    babelConfig('graph', moduleType)
+  );
+
+
```
Comment (lreeves) [inline build-lib/lib.js]: So much space for activities!

File: graph/fetch-schema.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import fetch from 'node-fetch';
+import introspectionQuery from './introspection-query';
+
+const baseGraphUrl = 'https://graphql.myshopify.com/api/graph';
+
+export default function fetchSchema() {
+  return fetch(baseGraphUrl, {
+    method: 'POST',
+    headers: {
+      Authorization: 'Basic MzUxYzEyMjAxN2QwZjJhOTU3ZDMyYWU3MjhhZDc0OWM=',
```
Comment (tessalt) [inline graph/fetch-schema.js]: we should probably put this somewhere else

File: graph/generator.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    /* eslint-disable no-underscore-dangle */
+    return extractTypeData(schema.data.__schema.types).filter(rejectBuiltins);
+    /* eslint-enable no-underscore-dangle */
+  };
+
+  if (cachedSchema) {
+    return Promise.resolve(cachedSchema).then(processSchema).then(exportTypesToFiles);
+  }
+
+  return fetchSchema().then(cacheSchema).then(processSchema).then(exportTypesToFiles);
```
Comment (tessalt) [inline graph/generator.js]: `catch`?

File: graph/fetch-schema.js (hunk 2 above)
Comment (minasmart) [inline graph/fetch-schema.js]: It's not a secret, so I'm not tooooo worried about it. Ideally I'd like to have a common config that devs can use where they can pull the schema from their own shop. Until then: recommendation for where this should go? Or OK here? (this branch is prototype-y)

File: graph/generator.js (hunk 3 above)
Comment (minasmart) [inline graph/generator.js]: None of the errors here are recoverable, so a catch would just be better error reporting. Is that what you're after?

Comment (tessalt) [inline graph/generator.js]: yep. having been bitten by useless promise errors lately.

Comment (minasmart) [inline graph/generator.js]: Cool. I'll add in some simple reporting.

**Comment by lreeves** [comment]
I'm probably not the right person to pour through the ES stuff but overall the GraphQL stuff looks great 👍

File: graph/fetch-schema.js (hunk 2 above)
Comment (tessalt) [inline graph/fetch-schema.js]: ok for now

**Pull request merged**